nx/lib/nx/defn/kernel.ex

```diff
@@ -1046,42 +1046,79 @@ defmodule Nx.Defn.Kernel do
 
   Similarly, to compute the factorial of `x` using `while`:
 
-        defn factorial(x) do
-          {factorial, _} =
-            while {factorial = 1, x}, Nx.greater(x, 1) do
-              {factorial * x, x - 1}
-            end
+      defn factorial(x) do
+        {factorial, _} =
+          while {factorial = 1, x}, Nx.greater(x, 1) do
+            {factorial * x, x - 1}
+          end
 
-          factorial
-        end
+        factorial
+      end
+
+  ## Generators
+
+  Inspired by Elixir's [for-comprehensions](`Kernel.SpecialForms.for/1`),
+  `while` in `defn` supports generators. For example, you could sum a one
+  dimensional tensor as follows:
+
+      while acc = 0, i <- tensor do
+        acc + i
+      end
+
+  You can give any non-scalar tensor and it will traverse the highest dimension
+  by default. One advantage of using generators is that you can also unroll the
+  loop for performance:
+
+      while acc = 0, i <- tensor, unroll: true do
+        acc + i
+      end
+
+  Or unroll it in batches:
+
+      while acc = 0, i <- tensor, unroll: 4 do
+        acc + i
+      end
 
```

Comment:
I think we should also add a warning note that `while` should generally be avoided unless there is no way to express the algorithm with regular tensor operations.

Reply:
Done!